File: site/src/chart/showcase.ts
// The config-centric reading of a showcase bot's curve: one run per span the
// bot ran one config, re-based to its own start. Derived here from the public
// artifact (points, switches, resets) so the collector publishes one shape and
// the config page and the bot page read the same file.

import type { ShowcaseBot } from "../data/static";
import { type BotReturnSeries, type ViewPoint } from "./returnCurve";

const DAY = 86400;

// The config page shows the bots that ran a config most recently when many did.
export const MAX_BOTS_PER_TEMPLATE = 5;

// At or below this the account was wiped out; a run cannot start from it.
const DEAD_EPS = 1e-9;

export type Run = {
  bot: { id: string; name: string; public_url: string };
  template_name: string;
  /** The capital the bot ran this span at, rounded to a magnitude by the collector. */
  cap_usdt: number;
  start: number;
  /** null while the span is still running. */
  end: number | null;
  /** Calendar days from the span's start to its end (or the collection time). */
  days: number;
  return_pct: number;
  points: ViewPoint[];
};

type Span = { start: number; end: number | null; cap_usdt: number; fromReset: boolean };

// The public file read as the owner's series shape, so the same windowing
// draws both: only the resets differ, a bare timestamp on the private side.
export function asSeries(bot: ShowcaseBot): BotReturnSeries {
  return { ...bot, capital_resets: (bot.capital_resets ?? []).map((r) => r.ts) };
}

// Every run of every config on one bot, newest first. A span opens at a
// config switch and closes at the next; a capital reset inside it splits it,
// since the index restarts at 100 there and nothing may be re-based across.
// A run counts by the calendar, as the exchange counts: the bot may have
// been idle on some of the days.
export function deriveRuns(bot: ShowcaseBot, now: number = bot.generated_at): Run[] {
  const points = (bot.points ?? []).slice().sort((a, b) => a.ts - b.ts);
  const switches = (bot.config_switches ?? []).slice().sort((a, b) => a.ts - b.ts);
  const resets = (bot.capital_resets ?? []).slice().sort((a, b) => a.ts - b.ts);
  const runs: Run[] = [];

  switches.forEach((sw, i) => {
    const spanEnd = switches[i + 1]?.ts ?? null;
    const inner = resets.filter((r) => r.ts > sw.ts && (spanEnd == null || r.ts < spanEnd));
    let span: Span = { start: sw.ts, end: inner[0]?.ts ?? spanEnd, cap_usdt: sw.cap_usdt, fromReset: false };
    for (let k = 0; ; k++) {
      const run = runOf(bot, points, sw.template_name, span, now);
      if (run) runs.push(run);
      const reset = inner[k];
      if (!reset) break;
      span = { start: reset.ts, end: inner[k + 1]?.ts ?? spanEnd, cap_usdt: reset.cap_usdt, fromReset: true };
    }
  });

  return runs.sort((a, b) => b.start - a.start);
}

function runOf(bot: ShowcaseBot, points: ShowcaseBot["points"], template: string, span: Span, now: number): Run | null {
  const endTs = span.end ?? now;
  const days = Math.floor((endTs - span.start) / DAY);
  const inside = points.filter((p) => p.ts >= span.start && p.ts <= endTs);
  // A span with no daily close yet has nothing to draw.
  if (inside.length === 0) return null;
  // The base is the close the span started from: the last point before a
  // switch, the reset day's own point after a re-funding (the day before held
  // the dust the deposit replaced).
  const before = span.fromReset ? undefined : points.filter((p) => p.ts < span.start).pop();
  const base = (before ?? inside[0]!).index;
  if (!(base > DEAD_EPS)) return null;
  const view: ViewPoint[] = inside.map((p) => ({ ts: p.ts, return_pct: (p.index / base - 1) * 100 }));
  return {
    bot: { id: bot.id, name: bot.name, public_url: bot.public_url },
    template_name: template,
    cap_usdt: span.cap_usdt,
    start: span.start,
    end: span.end,
    days,
    return_pct: view[view.length - 1]!.return_pct,
    points: view,
  };
}

/** Every run of one config on one bot, newest first. */
export type BotRuns = { bot: Run["bot"]; runs: Run[] };

// The showcase bots that ran one config, each with all its runs of it: a bot
// that left the config and came back is one entry, not several. Ordered by
// each bot's newest run, capped by bot.
export function botsForTemplate(bots: ShowcaseBot[], template: string, limit = MAX_BOTS_PER_TEMPLATE): BotRuns[] {
  return bots
    .map((b) => ({ runs: deriveRuns(b).filter((r) => r.template_name === template) }))
    .filter((g) => g.runs.length > 0)
    .map(({ runs }) => ({ bot: runs[0]!.bot, runs }))
    .sort((a, b) => b.runs[0]!.start - a.runs[0]!.start)
    .slice(0, limit);
}

// "$500", "$1k", "$1.5k", "$20k": the collector's rounded capital, in the
// magnitude it rounds to.
export function fmtCap(v: number): string {
  if (!(v > 0)) return "—";
  if (v >= 1_000_000) return `$${(v / 1_000_000).toFixed(v % 1_000_000 ? 1 : 0)}M`;
  if (v >= 1000) return `$${(v / 1000).toFixed(v % 1000 ? 1 : 0)}k`;
  return `$${v.toFixed(0)}`;
}

File: src/domain/showcase.rs
//! The public showcase: the artifact a bot on the operator's showcase is
//! published as, the listing built from those artifacts, and the port that
//! puts a bot on the page or takes it off.
//!
//! The artifact is a type of its own rather than the private series with
//! fields removed: it has no place for realized PnL or a balance, so a money
//! field added to the private series later cannot leak through it. `cap_usdt`
//! is the one balance-derived figure, rounded to a magnitude by the collector
//! that builds it.

use async_trait::async_trait;
use serde::{Deserialize, Serialize};

use crate::domain::bot::Bot;
use crate::domain::error::DomainError;

/// One day of a public curve: where the index stands, and nothing about money.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicPoint {
    pub ts: i64,
    pub index: f64,
    pub return_pct: f64,
}

/// A config switch on a public curve, with the capital the bot ran the new
/// config at: the wallet balance at the close of the last day before the
/// switch, rounded to a magnitude.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicSwitch {
    pub ts: i64,
    pub template_name: String,
    pub cap_usdt: f64,
}

/// A capital reset on a public curve, with the capital the new era started
/// at: the close of the reset day itself, since the day before held the dust
/// the re-funding replaced.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicReset {
    pub ts: i64,
    pub cap_usdt: f64,
}

/// The showcase artifact for one bot. `id` is the opaque public id, not the
/// bot id: bot ids are per-tenant keys two tenants may share.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicBotSeries {
    pub id: String,
    pub name: String,
    pub exchange: String,
    /// The copy-trading page the showcase links to; `None` on a bot shown
    /// without one.
    pub public_url: Option<String>,
    pub generated_at: i64,
    pub current_return_pct: f64,
    pub points: Vec<PublicPoint>,
    pub config_switches: Vec<PublicSwitch>,
    pub capital_resets: Vec<PublicReset>,
}

impl PublicBotSeries {
    /// Take the bot's current name and link. The curve may be a day old; the
    /// name and the link are the row's, so a rename or a cleared link since
    /// the curve was built is never published from the artifact.
    pub fn refresh_from(&mut self, bot: &Bot) {
        self.name = bot.name.clone();
        self.public_url = bot.public_url.clone();
    }
}

/// The showcase listing: every public bot with what a list row needs, and a
/// 30-day sparkline. Written whenever the published set changes, empty when
/// nothing is public, so the page can tell "nothing public" from "never
/// written".
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicIndex {
    pub generated_at: i64,
    pub bots: Vec<PublicIndexBot>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PublicIndexBot {
    pub id: String,
    pub name: String,
    pub exchange: String,
    pub public_url: Option<String>,
    pub current_return_pct: f64,
    pub spark: Vec<f64>,
}

const SPARK_DAYS: usize = 30;

impl PublicIndex {
    pub fn new(series: &[PublicBotSeries], generated_at: i64) -> Self {
        let bots = series
            .iter()
            .map(|s| PublicIndexBot {
                id: s.id.clone(),
                name: s.name.clone(),
                exchange: s.exchange.clone(),
                public_url: s.public_url.clone(),
                current_return_pct: s.current_return_pct,
                spark: s
                    .points
                    .iter()
                    .rev()
                    .take(SPARK_DAYS)
                    .map(|p| p.return_pct)
                    .collect::<Vec<_>>()
                    .into_iter()
                    .rev()
                    .collect(),
            })
            .collect();
        Self { generated_at, bots }
    }
}

/// `sha256("{user_id}#{bot_id}")`, twelve hex characters: stable for the life
/// of the bot, distinct across tenants, and safe as an S3 key and a URL
/// segment whatever the bot was named.
pub fn public_id(user_id: &str, bot_id: &str) -> String {
    use sha2::{Digest, Sha256};
    hex::encode(&Sha256::digest(format!("{user_id}#{bot_id}"))[..6])
}

/// Whether a bot is published: it is on the showcase, on the operator's
/// account. The same choice on any other account's row is inert.
pub fn is_published(bot: &Bot, operator: bool) -> bool {
    operator && bot.on_showcase()
}

/// What a publish found for the bot.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Published {
    /// The bot's artifact is on the public prefix.
    Live,
    /// No curve has been collected for the bot yet (no stored keys, or a bot
    /// added since the last run), so
    /// there is nothing to put on the page.
    NoCurveYet,
}

/// Puts a bot on the public showcase page or takes it off, and keeps the
/// listing in step with what is published. The caller has already decided the
/// bot belongs there; this only moves artifacts.
#[async_trait]
pub trait ShowcasePublisher: Send + Sync {
    async fn publish(&self, bot: &Bot) -> Result<Published, DomainError>;
    async fn withdraw(&self, bot: &Bot) -> Result<(), DomainError>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::exchange::Exchange;

    const LINK: &str = "https://www.bybit.com/copyTrade/x";

    fn a_public_bot() -> Bot {
        let mut bot = Bot::create(
            "u-1".into(),
            Exchange::Bybit,
            "shown".into(),
            "ak".into(),
            "sk".into(),
            1,
        );
        bot.set_public_url(Some(LINK.into()), 1).unwrap();
        bot
    }

    fn series_with(id: &str, days: usize) -> PublicBotSeries {
        PublicBotSeries {
            id: id.into(),
            name: "old name".into(),
            exchange: "bybit".into(),
            public_url: Some(LINK.into()),
            generated_at: 0,
            current_return_pct: days as f64,
            points: (0..days)
                .map(|d| PublicPoint {
                    ts: d as i64 * 86_400,
                    index: 100.0 + d as f64,
                    return_pct: d as f64,
                })
                .collect(),
            config_switches: vec![PublicSwitch {
                ts: 0,
                template_name: "tpl-a".into(),
                cap_usdt: 700.0,
            }],
            capital_resets: vec![],
        }
    }

    #[test]
    fn public_id_is_twelve_hex_of_user_and_bot() {
        let id = public_id("u-1", "shown");
        assert_eq!(id.len(), 12);
        assert!(id.bytes().all(|b| b.is_ascii_hexdigit()));
        assert_eq!(id, public_id("u-1", "shown"), "stable");
        assert_ne!(id, public_id("u-2", "shown"), "per tenant");
    }

    #[test]
    fn a_bot_is_published_when_it_is_on_the_operators_showcase() {
        let private = Bot::create(
            "u-1".into(),
            Exchange::Bybit,
            "quiet".into(),
            "ak".into(),
            "sk".into(),
            1,
        );
        assert!(!is_published(&private, true));
        assert!(is_published(&a_public_bot(), true));
        assert!(!is_published(&a_public_bot(), false), "not an operator");

        let mut hidden = a_public_bot();
        hidden.set_showcase(false, 2);
        assert!(
            !is_published(&hidden, true),
            "hidden keeps the link, not the page"
        );

        let mut unlinked = Bot::create(
            "u-1".into(),
            Exchange::Bybit,
            "bare".into(),
            "ak".into(),
            "sk".into(),
            1,
        );
        unlinked.set_showcase(true, 2);
        assert!(is_published(&unlinked, true));
    }

    #[test]
    fn the_index_lists_each_artifact_with_its_last_thirty_days_oldest_first() {
        let index = PublicIndex::new(&[series_with("a", 40), series_with("b", 3)], 7);
        assert_eq!(index.generated_at, 7);
        assert_eq!(
            index.bots[0].spark,
            (10..40).map(|d| d as f64).collect::<Vec<_>>()
        );
        assert_eq!(index.bots[1].spark, vec![0.0, 1.0, 2.0]);
        assert_eq!(index.bots[1].id, "b");
    }

    #[test]
    fn refreshing_takes_the_rows_name_and_link_and_keeps_the_curve() {
        let mut series = series_with("a", 3);
        let mut bot = Bot::create(
            "u-1".into(),
            Exchange::Bybit,
            "renamed".into(),
            "ak".into(),
            "sk".into(),
            1,
        );
        bot.set_showcase(true, 2);
        series.refresh_from(&bot);
        assert_eq!(series.name, "renamed");
        assert_eq!(series.public_url, None, "a cleared link is not republished");
        assert_eq!(series.points.len(), 3);
    }

    #[test]
    fn an_artifact_reads_back_as_it_was_written() {
        let series = series_with("a", 2);
        let read: PublicBotSeries =
            serde_json::from_slice(&serde_json::to_vec(&series).unwrap()).unwrap();
        assert_eq!(read, series);
    }
}
